File: python/api/providers/google.py
"""
Google Gemini provider adapter.

Converts between OpenAI and Google Gemini API formats.
"""

from __future__ import annotations

import json
import logging
import time
import uuid
from typing import Any, AsyncIterator

import httpx

logger = logging.getLogger(__name__)

DEFAULT_TIMEOUT = 120.0
MAX_RETRIES = 2
RETRY_DELAY = 0.5

# Map common model aliases to Gemini model IDs.
MODEL_MAP: dict[str, str] = {
    "gemini-pro": "gemini-1.5-pro",
    "gemini-1.5-pro": "gemini-1.5-pro-latest",
    "gemini-1.5-flash": "gemini-1.5-flash-latest",
    "gemini-2.0-flash": "gemini-2.0-flash",
    "gemini-2.5-pro": "gemini-2.5-pro-preview-05-06",
    "gemini-2.5-flash": "gemini-2.5-flash-preview-05-20",
}


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

async def chat_completion(
    request_data: dict[str, Any],
    provider_config: dict[str, Any],
) -> dict[str, Any] | AsyncIterator[bytes]:
    """Send a chat completion request to Google Gemini.

    Parameters
    ----------
    request_data:
        OpenAI-format request body.
    provider_config:
        {
            "base_url": "https://generativelanguage.googleapis.com",  # optional
            "api_key": "AIza...",
            "extra_headers": {},
            "timeout": 120,
        }
    """
    base_url = provider_config.get(
        "base_url", "https://generativelanguage.googleapis.com"
    ).rstrip("/")
    api_key = provider_config["api_key"]
    timeout = provider_config.get("timeout", DEFAULT_TIMEOUT)
    extra_headers = provider_config.get("extra_headers", {})
    is_stream = request_data.get("stream", False)

    gemini_body = _to_gemini_request(request_data)
    model = gemini_body.get("_model", "gemini-1.5-pro-latest")
    del gemini_body["_model"]

    if is_stream:
        endpoint = f"{base_url}/v1beta/models/{model}:streamGenerateContent?alt=sse&key={api_key}"
    else:
        endpoint = f"{base_url}/v1beta/models/{model}:generateContent?key={api_key}"

    headers = {"Content-Type": "application/json", **extra_headers}

    return await _do_request(
        endpoint, headers, gemini_body, timeout, is_stream,
        request_data.get("model", model),
    )


# ---------------------------------------------------------------------------
# Format conversion: OpenAI -> Gemini
# ---------------------------------------------------------------------------

def _to_gemini_request(openai_req: dict[str, Any]) -> dict[str, Any]:
    """Convert OpenAI-format request to Gemini generateContent format."""

    model = openai_req.get("model", "gemini-1.5-pro")
    mapped_model = MODEL_MAP.get(model, model)

    messages_raw: list[dict[str, Any]] = openai_req.get("messages", [])

    system_instruction: dict[str, Any] | None = None
    contents: list[dict[str, Any]] = []

    for msg in messages_raw:
        role = msg.get("role", "user")
        content = msg.get("content", "")

        if role == "system":
            system_instruction = _build_content_part(content)
            continue

        gemini_role = "user" if role in ("user", "tool") else "model"
        parts = _build_parts(content)
        contents.append({"role": gemini_role, "parts": parts})

    # Gemini requires alternating user/model roles.
    # If the first message isn't from user, insert a placeholder.
    if contents and contents[0]["role"] != "user":
        contents.insert(0, {"role": "user", "parts": [{"text": ""}]})

    # Ensure alternating roles: merge consecutive same-role messages.
    contents = _merge_consecutive(contents)

    body: dict[str, Any] = {
        "_model": mapped_model,
        "contents": contents,
    }

    if system_instruction:
        body["systemInstruction"] = system_instruction

    # Generation config
    gen_config: dict[str, Any] = {}
    if "temperature" in openai_req:
        gen_config["temperature"] = openai_req["temperature"]
    if "top_p" in openai_req:
        gen_config["topP"] = openai_req["top_p"]
    if "max_tokens" in openai_req:
        gen_config["maxOutputTokens"] = openai_req["max_tokens"]
    if "stop" in openai_req:
        stop = openai_req["stop"]
        if isinstance(stop, str):
            stop = [stop]
        gen_config["stopSequences"] = stop
    if "presence_penalty" in openai_req:
        gen_config["presencePenalty"] = openai_req["presence_penalty"]
    if "frequency_penalty" in openai_req:
        gen_config["frequencyPenalty"] = openai_req["frequency_penalty"]

    if gen_config:
        body["generationConfig"] = gen_config

    return body


def _build_parts(content: Any) -> list[dict[str, Any]]:
    """Convert OpenAI content to Gemini parts."""
    if isinstance(content, str):
        return [{"text": content}]
    if isinstance(content, list):
        parts: list[dict[str, Any]] = []
        for part in content:
            ptype = part.get("type", "text")
            if ptype == "text":
                parts.append({"text": part.get("text", "")})
            elif ptype == "image_url":
                url = part.get("image_url", {}).get("url", "")
                if url.startswith("data:"):
                    # data:image/png;base64,xxxx
                    header, b64data = url.split(",", 1)
                    mime = header.split(";")[0].split(":")[1] if ":" in header else "image/png"
                    parts.append({
                        "inlineData": {
                            "mimeType": mime,
                            "data": b64data,
                        }
                    })
        return parts if parts else [{"text": ""}]
    return [{"text": str(content)}]


def _build_content_part(content: Any) -> dict[str, Any]:
    """Build a system instruction content block."""
    if isinstance(content, str):
        return {"parts": [{"text": content}]}
    if isinstance(content, list):
        texts = []
        for part in content:
            if part.get("type") == "text":
                texts.append(part.get("text", ""))
        return {"parts": [{"text": "\n".join(texts)}]}
    return {"parts": [{"text": str(content)}]}


def _merge_consecutive(contents: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Merge consecutive messages with the same role."""
    merged: list[dict[str, Any]] = []
    for msg in contents:
        if merged and merged[-1]["role"] == msg["role"]:
            # Merge parts
            merged[-1]["parts"].extend(msg["parts"])
        else:
            merged.append({"role": msg["role"], "parts": list(msg["parts"])})
    return merged


# ---------------------------------------------------------------------------
# Format conversion: Gemini -> OpenAI
# ---------------------------------------------------------------------------

def _to_openai_response(
    gemini_resp: dict[str, Any],
    original_model: str,
) -> dict[str, Any]:
    """Convert Gemini generateContent response to OpenAI format."""

    candidates = gemini_resp.get("candidates", [])
    text = ""
    finish_reason = "stop"

    if candidates:
        candidate = candidates[0]
        content = candidate.get("content", {})
        parts = content.get("parts", [])
        text_parts = [p.get("text", "") for p in parts if "text" in p]
        text = "".join(text_parts)

        reason = candidate.get("finishReason", "STOP")
        if reason == "MAX_TOKENS":
            finish_reason = "length"
        elif reason == "SAFETY":
            finish_reason = "content_filter"

    usage_meta = gemini_resp.get("usageMetadata", {})
    prompt_tokens = usage_meta.get("promptTokenCount", 0)
    completion_tokens = usage_meta.get("candidatesTokenCount", 0)

    return {
        "id": f"chatcmpl-{uuid.uuid4().hex[:24]}",
        "object": "chat.completion",
        "created": int(time.time()),
        "model": original_model,
        "choices": [
            {
                "index": 0,
                "message": {
                    "role": "assistant",
                    "content": text,
                },
                "finish_reason": finish_reason,
            }
        ],
        "usage": {
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "total_tokens": prompt_tokens + completion_tokens,
        },
    }


# ---------------------------------------------------------------------------
# HTTP request with retry
# ---------------------------------------------------------------------------

async def _do_request(
    url: str,
    headers: dict[str, str],
    body: dict[str, Any],
    timeout: float,
    is_stream: bool,
    original_model: str,
) -> dict[str, Any] | AsyncIterator[bytes]:

    last_exc: Exception | None = None

    for attempt in range(MAX_RETRIES + 1):
        try:
            if is_stream:
                return _stream_response(url, headers, body, timeout, original_model)

            async with httpx.AsyncClient(timeout=timeout) as client:
                resp = await client.post(url, headers=headers, json=body)
                resp.raise_for_status()
                data = resp.json()
                return _to_openai_response(data, original_model)

        except (httpx.HTTPStatusError, httpx.ConnectError, httpx.ReadTimeout) as exc:
            last_exc = exc
            status = getattr(getattr(exc, "response", None), "status_code", None)

            if status is not None and 400 <= status < 500 and status != 429:
                raise _wrap_error(exc) from exc

            if attempt < MAX_RETRIES:
                logger.warning(
                    "Gemini request failed (attempt %d/%d): %s – retrying",
                    attempt + 1,
                    MAX_RETRIES + 1,
                    exc,
                )
                time.sleep(RETRY_DELAY * (attempt + 1))
                continue

            raise _wrap_error(exc) from exc

    raise _wrap_error(last_exc) from last_exc  # type: ignore[arg-type]


async def _stream_response(
    url: str,
    headers: dict[str, str],
    body: dict[str, Any],
    timeout: float,
    original_model: str,
) -> AsyncIterator[bytes]:
    """Convert Gemini SSE stream to OpenAI SSE stream."""

    completion_id = f"chatcmpl-{uuid.uuid4().hex[:24]}"
    created = int(time.time())

    async with httpx.AsyncClient(timeout=timeout) as client:
        async with client.stream("POST", url, headers=headers, json=body) as resp:
            resp.raise_for_status()

            async for line in resp.aiter_lines():
                if not line.startswith("data: "):
                    continue
                payload = line[len("data: "):]
                if not payload.strip():
                    continue

                try:
                    event = json.loads(payload)
                except json.JSONDecodeError:
                    continue

                # Check for errors
                if "error" in event:
                    err = event["error"]
                    raise Exception(
                        f"Gemini stream error: {err.get('code', 'unknown')}: "
                        f"{err.get('message', 'unknown error')}"
                    )

                candidates = event.get("candidates", [])
                if not candidates:
                    continue

                candidate = candidates[0]
                content = candidate.get("content", {})
                parts = content.get("parts", [])
                text = "".join(p.get("text", "") for p in parts if "text" in p)

                finish_reason = None
                reason = candidate.get("finishReason")
                if reason:
                    finish_reason = "length" if reason == "MAX_TOKENS" else "stop"

                chunk = {
                    "id": completion_id,
                    "object": "chat.completion.chunk",
                    "created": created,
                    "model": original_model,
                    "choices": [
                        {
                            "index": 0,
                            "delta": {"content": text} if text else {},
                            "finish_reason": finish_reason,
                        }
                    ],
                }

                # Include usage in final chunk if present
                usage_meta = event.get("usageMetadata")
                if usage_meta and finish_reason:
                    chunk["usage"] = {
                        "prompt_tokens": usage_meta.get("promptTokenCount", 0),
                        "completion_tokens": usage_meta.get("candidatesTokenCount", 0),
                        "total_tokens": usage_meta.get("totalTokenCount", 0),
                    }

                yield f"data: {json.dumps(chunk)}\n\n".encode("utf-8")

    yield b"data: [DONE]\n\n"


# ---------------------------------------------------------------------------
# Usage extraction
# ---------------------------------------------------------------------------

def extract_usage(response_data: dict[str, Any]) -> dict[str, int]:
    """Extract token usage from an already-converted OpenAI-format response."""
    usage = response_data.get("usage", {})
    return {
        "input_tokens": usage.get("prompt_tokens", 0),
        "output_tokens": usage.get("completion_tokens", 0),
    }


# ---------------------------------------------------------------------------
# Error helpers
# ---------------------------------------------------------------------------

def _wrap_error(exc: Exception) -> Exception:
    if isinstance(exc, httpx.HTTPStatusError):
        body = ""
        try:
            body = exc.response.text
        except Exception:  # noqa: BLE001
            pass
        return Exception(f"Gemini API error {exc.response.status_code}: {body}")
    if isinstance(exc, httpx.ReadTimeout):
        return Exception("Gemini API request timed out")
    return exc

File: nodejs/src/api/providers/google.ts
/**
 * Google Gemini provider adapter.
 *
 * Converts between OpenAI and Google Gemini API formats.
 */

import { v4 as uuidv4 } from "uuid";
import { injectForGoogle, type ThinkingLevel } from "../thinkingParser.js";
import { createRequestTimeout, isAbortError, withRequestTimeout } from "./requestTimeout.js";
import { ProviderHttpError } from "./errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ProviderConfig {
  baseUrl?: string;
  apiKey: string;
  extraHeaders?: Record<string, string>;
  timeout?: number;
}

export interface ChatCompletionRequest {
  model: string;
  messages: Array<{
    role: string;
    content: string | any[];
  }>;
  stream?: boolean;
  temperature?: number;
  top_p?: number;
  max_tokens?: number;
  stop?: string | string[];
  presence_penalty?: number;
  frequency_penalty?: number;
  [key: string]: unknown;
}

export interface Usage {
  input_tokens: number;
  output_tokens: number;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_TIMEOUT = 120_000;
const MAX_RETRIES = 2;
const RETRY_DELAY = 500;

const MODEL_MAP: Record<string, string> = {
  "gemini-pro": "gemini-1.5-pro",
  "gemini-1.5-pro": "gemini-1.5-pro-latest",
  "gemini-1.5-flash": "gemini-1.5-flash-latest",
  "gemini-2.0-flash": "gemini-2.0-flash",
  "gemini-2.5-pro": "gemini-2.5-pro-preview-05-06",
  "gemini-2.5-flash": "gemini-2.5-flash-preview-05-20",
};

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export async function chatCompletion(
  requestData: ChatCompletionRequest,
  providerConfig: ProviderConfig
): Promise<Record<string, unknown> | AsyncGenerator<Uint8Array>> {
  const baseUrl = (
    providerConfig.baseUrl ?? "https://generativelanguage.googleapis.com"
  ).replace(/\/+$/, "");
  const apiKey = providerConfig.apiKey;
  const timeout = providerConfig.timeout ?? DEFAULT_TIMEOUT;
  const extraHeaders = providerConfig.extraHeaders ?? {};
  const isStream = requestData.stream === true;

  const geminiBody = toGeminiRequest(requestData);
  const model = geminiBody._model as string;
  // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
  delete (geminiBody as Record<string, unknown>)._model;

  const endpoint = isStream
    ? `${baseUrl}/v1beta/models/${model}:streamGenerateContent?alt=sse&key=${apiKey}`
    : `${baseUrl}/v1beta/models/${model}:generateContent?key=${apiKey}`;

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    ...extraHeaders,
  };

  return doRequest(endpoint, headers, geminiBody, timeout, isStream, requestData.model);
}

// ---------------------------------------------------------------------------
// Format conversion: OpenAI -> Gemini
// ---------------------------------------------------------------------------

function toGeminiRequest(
  openaiReq: ChatCompletionRequest
): Record<string, any> & { _model: string } {
  const model = openaiReq.model ?? "gemini-1.5-pro";
  const mappedModel = MODEL_MAP[model] ?? model;

  const messagesRaw = openaiReq.messages ?? [];

  const systemInstructionParts: Array<Record<string, any>> = [];
  const contents: Array<Record<string, any>> = [];

  for (const msg of messagesRaw) {
    const role = msg.role ?? "user";
    const content = msg.content;

    if (role === "system") {
      systemInstructionParts.push(...buildParts(content));
      continue;
    }

    const geminiRole = role === "user" || role === "tool" ? "user" : "model";
    const parts = buildParts(content);
    contents.push({ role: geminiRole, parts });
  }

  // Gemini requires first message to be from user
  if (contents.length > 0 && contents[0].role !== "user") {
    contents.unshift({ role: "user", parts: [{ text: "" }] });
  }

  // Ensure alternating roles
  const merged = mergeConsecutive(contents);

  const body: Record<string, any> & { _model: string } = {
    _model: mappedModel,
    contents: merged,
  };

  if (systemInstructionParts.length > 0) {
    body.systemInstruction = { parts: systemInstructionParts };
  }

  // Generation config
  const genConfig: Record<string, any> = {};
  if (openaiReq.temperature !== undefined) genConfig.temperature = openaiReq.temperature;
  if (openaiReq.top_p !== undefined) genConfig.topP = openaiReq.top_p;
  if (openaiReq.max_tokens !== undefined) genConfig.maxOutputTokens = openaiReq.max_tokens;
  if (openaiReq.stop !== undefined) {
    genConfig.stopSequences = Array.isArray(openaiReq.stop)
      ? openaiReq.stop
      : [openaiReq.stop];
  }
  if (openaiReq.presence_penalty !== undefined)
    genConfig.presencePenalty = openaiReq.presence_penalty;
  if (openaiReq.frequency_penalty !== undefined)
    genConfig.frequencyPenalty = openaiReq.frequency_penalty;

  if (Object.keys(genConfig).length > 0) {
    body.generationConfig = genConfig;
  }

  // --- thinking budget (Gemini thinkingConfig) ---
  if (openaiReq.thinking_effort) {
    injectForGoogle(body, openaiReq.thinking_effort as ThinkingLevel);
  }

  return body;
}

function buildParts(content: any): Array<Record<string, any>> {
  if (typeof content === "string") return [{ text: content }];
  if (Array.isArray(content)) {
    const parts: Array<Record<string, any>> = [];
    for (const part of content) {
      const ptype = part.type ?? "text";
      if (ptype === "text") {
        parts.push({ text: part.text ?? "" });
      } else if (ptype === "image_url") {
        const url: string = part.image_url?.url ?? "";
        if (url.startsWith("data:")) {
          const [header, b64data] = url.split(",", 2);
          const mime = header.includes(":")
            ? header.split(";")[0].split(":")[1]
            : "image/png";
          parts.push({ inlineData: { mimeType: mime, data: b64data } });
        }
      }
    }
    return parts.length > 0 ? parts : [{ text: "" }];
  }
  return [{ text: String(content) }];
}

function mergeConsecutive(
  contents: Array<Record<string, any>>
): Array<Record<string, any>> {
  const merged: Array<Record<string, any>> = [];
  for (const msg of contents) {
    if (merged.length > 0 && merged[merged.length - 1].role === msg.role) {
      merged[merged.length - 1].parts.push(...msg.parts);
    } else {
      merged.push({ role: msg.role, parts: [...msg.parts] });
    }
  }
  return merged;
}

// ---------------------------------------------------------------------------
// Format conversion: Gemini -> OpenAI
// ---------------------------------------------------------------------------

function toOpenAIResponse(
  geminiResp: Record<string, any>,
  originalModel: string
): Record<string, any> {
  const candidates = geminiResp.candidates ?? [];
  let text = "";
  let finishReason = "stop";

  if (candidates.length > 0) {
    const candidate = candidates[0];
    const parts = candidate.content?.parts ?? [];
    text = parts.filter((p: any) => "text" in p).map((p: any) => p.text).join("");

    finishReason = mapFinishReason(candidate.finishReason ?? "STOP") ?? "stop";
  }

  const usageMeta = geminiResp.usageMetadata ?? {};
  const promptTokens = usageMeta.promptTokenCount ?? 0;
  const completionTokens = usageMeta.candidatesTokenCount ?? 0;

  return {
    id: `chatcmpl-${uuidv4().replace(/-/g, "").slice(0, 24)}`,
    object: "chat.completion",
    created: Math.floor(Date.now() / 1000),
    model: originalModel,
    choices: [
      {
        index: 0,
        message: { role: "assistant", content: text },
        finish_reason: finishReason,
      },
    ],
    usage: {
      prompt_tokens: promptTokens,
      completion_tokens: completionTokens,
      total_tokens: promptTokens + completionTokens,
    },
  };
}

// ---------------------------------------------------------------------------
// HTTP request with retry
// ---------------------------------------------------------------------------

async function doRequest(
  url: string,
  headers: Record<string, string>,
  body: Record<string, any>,
  timeout: number,
  isStream: boolean,
  originalModel: string
): Promise<Record<string, unknown> | AsyncGenerator<Uint8Array>> {
  let lastError: Error | null = null;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      if (isStream) {
        return streamResponse(url, headers, body, timeout, originalModel);
      }

      return await withRequestTimeout(timeout, async (signal) => {
        const resp = await fetch(url, {
          method: "POST",
          headers,
          body: JSON.stringify(body),
          signal,
        });

        if (!resp.ok) {
          const errorBody = await resp.text();
          throw new ProviderHttpError(`Gemini API error ${resp.status}: ${errorBody}`, resp.status);
        }

        const data = await resp.json();
        return toOpenAIResponse(data, originalModel);
      });
    } catch (err: any) {
      lastError = err;
      const status: number | undefined = err.status;

      if (status && status >= 400 && status < 500 && status !== 429) {
        throw wrapError(err);
      }

      if (attempt < MAX_RETRIES) {
        console.warn(
          `Gemini request failed (attempt ${attempt + 1}/${MAX_RETRIES + 1}): ${err} – retrying`
        );
        await sleep(RETRY_DELAY * (attempt + 1));
        continue;
      }

      throw wrapError(err);
    }
  }

  throw wrapError(lastError!);
}

// ---------------------------------------------------------------------------
// Streaming: Gemini SSE -> OpenAI SSE
// ---------------------------------------------------------------------------

async function* streamResponse(
  url: string,
  headers: Record<string, string>,
  body: Record<string, any>,
  timeout: number,
  originalModel: string
): AsyncGenerator<Uint8Array> {
  const completionId = `chatcmpl-${uuidv4().replace(/-/g, "").slice(0, 24)}`;
  const created = Math.floor(Date.now() / 1000);
  const encoder = new TextEncoder();

  const requestTimeout = createRequestTimeout(timeout);
  let reader: ReadableStreamDefaultReader<Uint8Array> | undefined;
  let completed = false;

  const decoder = new TextDecoder();
  let buffer = "";

  try {
    const resp = await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(body),
      signal: requestTimeout.signal,
    });

    if (!resp.ok) {
      const errorBody = await resp.text();
      throw new Error(`Gemini API error ${resp.status}: ${errorBody}`);
    }

    reader = resp.body?.getReader();
    if (!reader) throw new Error("No response body for streaming");

    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        completed = true;
        break;
      }

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed.startsWith("data: ")) continue;
        const payload = trimmed.slice(6).trim();
        if (!payload) continue;

        let event: Record<string, any>;
        try {
          event = JSON.parse(payload);
        } catch {
          continue;
        }

        // Check for errors
        if (event.error) {
          const err = event.error;
          throw new Error(
            `Gemini stream error: ${err.code ?? "unknown"}: ${err.message ?? "unknown error"}`
          );
        }

        const candidates = event.candidates ?? [];
        if (candidates.length === 0) continue;

        const candidate = candidates[0];
        const parts = candidate.content?.parts ?? [];
        const text = parts
          .filter((p: any) => "text" in p)
          .map((p: any) => p.text)
          .join("");

        let finishReason: string | null = null;
        const reason = candidate.finishReason;
        if (reason) {
          finishReason = mapFinishReason(reason);
        }

        const chunk: Record<string, any> = {
          id: completionId,
          object: "chat.completion.chunk",
          created,
          model: originalModel,
          choices: [
            {
              index: 0,
              delta: text ? { content: text } : {},
              finish_reason: finishReason,
            },
          ],
        };

        const usageMeta = event.usageMetadata;
        if (usageMeta && finishReason) {
          chunk.usage = {
            prompt_tokens: usageMeta.promptTokenCount ?? 0,
            completion_tokens: usageMeta.candidatesTokenCount ?? 0,
            total_tokens: usageMeta.totalTokenCount ?? 0,
          };
        }

        yield encoder.encode(`data: ${JSON.stringify(chunk)}\n\n`);
      }
    }
  } catch (err) {
    if (isAbortError(err) || requestTimeout.signal.aborted) {
      throw new Error("Gemini API request timed out");
    }
    throw err;
  } finally {
    requestTimeout.clear();
    if (reader) {
      if (!completed) {
        requestTimeout.abort();
        await reader.cancel().catch(() => undefined);
      }
      reader.releaseLock();
    }
  }

  yield encoder.encode("data: [DONE]\n\n");
}

function mapFinishReason(reason: string): string | null {
  if (reason === "MAX_TOKENS") return "length";
  if (reason === "SAFETY") return "content_filter";
  if (reason === "STOP") return "stop";
  return "stop";
}

// ---------------------------------------------------------------------------
// Usage extraction
// ---------------------------------------------------------------------------

export function extractUsage(responseData: Record<string, any>): Usage {
  const usage = responseData.usage ?? {};
  return {
    input_tokens: usage.prompt_tokens ?? 0,
    output_tokens: usage.completion_tokens ?? 0,
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function wrapError(err: Error): Error {
  if (err.name === "AbortError") {
    return new Error("Gemini API request timed out");
  }
  return err;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
